File: app-node/src/tools.ts
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { VectorStore } from "@langchain/core/vectorstores";

export function buildRetrieverTool(vectorStore: VectorStore, pdfName: string) {
    // k=3 similar to python
    const retriever = vectorStore.asRetriever({ k: 3 });

    // Sanitize name for tool
    const safeName = "search_" + pdfName.replace(".pdf", "").replace(/\s+/g, "_").replace(/-/g, "_").toLowerCase();

    return tool(
        async ({ query }) => {
            const docs = await retriever.invoke(query);
            // docs is an array of Documents
            if (!docs || docs.length === 0) return "No relevant information found.";

            return docs.map((doc, i) => `Source ${i + 1}:\n${doc.pageContent}`).join("\n\n");
        },
        {
            name: safeName,
            description: `Search and retrieve information from ${pdfName}`,
            schema: z.object({
                query: z.string().describe("The query to search the PDF for")
            })
        }
    );
}

File: app/tools.py
from langchain_core.tools import tool

def build_retriever_tool(retriever, pdf_name: str):
    """
    Build a retriever tool for a specific PDF
    
    Args:
        retriever: The vectorstore retriever
        pdf_name: Name of the PDF being queried
    """

    @tool
    def retriever_tool(query: str) -> str:
        """Search and retrieve information from the PDF"""
        docs = retriever.invoke(query)

        if not docs:
            return "No relevant information found."

        return "\n\n".join(
            f"Source {i+1}:\n{doc.page_content}"
            for i, doc in enumerate(docs)
        )
    
    # Update the tool's name and description dynamically
    retriever_tool.name = f"search_{pdf_name.replace('.pdf', '').replace(' ', '_').replace('-', '_').lower()}"
    retriever_tool.description = f"Search and retrieve information from {pdf_name}"

    return retriever_tool
